File: scripts/ollama-client-local.js
/**
 * Ollama Client for Local Computer (Node.js)
 * আপনার local computer থেকে AWS Ollama ব্যবহার করার জন্য
 */

const axios = require('axios');

class OllamaClient {
    constructor(baseUrl) {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.apiUrl = `${this.baseUrl}/api`;
    }

    /**
     * Server connection check করুন
     */
    async checkConnection() {
        try {
            const response = await axios.get(`${this.apiUrl}/tags`, { timeout: 5000 });
            return response.status === 200;
        } catch (error) {
            return false;
        }
    }

    /**
     * Available models-র list পান
     */
    async listModels() {
        try {
            const response = await axios.get(`${this.apiUrl}/tags`);
            return response.data.models.map(model => model.name);
        } catch (error) {
            console.error('Error:', error.message);
            return [];
        }
    }

    /**
     * Text generate করুন
     */
    async generate(prompt, model = 'llama3.2', options = {}) {
        try {
            const response = await axios.post(
                `${this.apiUrl}/generate`,
                {
                    model: model,
                    prompt: prompt,
                    stream: false,
                    options: {
                        temperature: options.temperature || 0.7
                    }
                },
                { timeout: 300000 } // 5 minutes
            );

            return response.data.response;
        } catch (error) {
            throw new Error(`Generation failed: ${error.message}`);
        }
    }

    /**
     * Chat with conversation history
     */
    async chat(messages, model = 'llama3.2') {
        try {
            const response = await axios.post(
                `${this.apiUrl}/chat`,
                {
                    model: model,
                    messages: messages,
                    stream: false
                },
                { timeout: 300000 }
            );

            return response.data.message.content;
        } catch (error) {
            throw new Error(`Chat failed: ${error.message}`);
        }
    }

    /**
     * Streaming generation
     */
    async *generateStream(prompt, model = 'llama3.2') {
        try {
            const response = await axios.post(
                `${this.apiUrl}/generate`,
                {
                    model: model,
                    prompt: prompt,
                    stream: true
                },
                {
                    responseType: 'stream',
                    timeout: 300000
                }
            );

            let buffer = '';

            for await (const chunk of response.data) {
                buffer += chunk.toString();
                const lines = buffer.split('\n');
                buffer = lines.pop() || '';

                for (const line of lines) {
                    if (line.trim()) {
                        try {
                            const data = JSON.parse(line);
                            if (data.response) {
                                yield data.response;
                            }
                        } catch (e) {
                            // Skip invalid JSON
                        }
                    }
                }
            }
        } catch (error) {
            throw new Error(`Streaming failed: ${error.message}`);
        }
    }
}

// Example usage
async function main() {
    // আপনার Ollama server-র URL দিন
    const OLLAMA_URL = 'http://54.123.45.67'; // আপনার EC2 IP দিন

    // Client initialize করুন
    const client = new OllamaClient(OLLAMA_URL);

    // Connection test করুন
    console.log('🔍 Testing connection...');
    const isConnected = await client.checkConnection();

    if (isConnected) {
        console.log('✅ Connected to Ollama server!');
    } else {
        console.log('❌ Cannot connect to Ollama server');
        process.exit(1);
    }

    // Available models দেখুন
    console.log('\n📋 Available models:');
    const models = await client.listModels();
    models.forEach(model => console.log(`  - ${model}`));

    // Simple generation test
    console.log('\n💬 Testing text generation...');
    const response = await client.generate(
        'What is the capital of Bangladesh?',
        'llama3.2'
    );
    console.log(`Response: ${response}`);

    // Chat test
    console.log('\n💬 Testing chat...');
    const messages = [
        { role: 'user', content: 'What is JavaScript?' },
        { role: 'assistant', content: 'JavaScript is a programming language.' },
        { role: 'user', content: 'What can I do with it?' }
    ];
    const chatResponse = await client.chat(messages, 'llama3.2');
    console.log(`Response: ${chatResponse}`);

    // Streaming test
    console.log('\n💬 Testing streaming...');
    process.stdout.write('Response: ');
    for await (const chunk of client.generateStream('Count from 1 to 5', 'llama3.2')) {
        process.stdout.write(chunk);
    }
    console.log('\n');

    console.log('✅ All tests completed!');
}

// Run if called directly
if (require.main === module) {
    main().catch(console.error);
}

module.exports = OllamaClient;

File: py/ollama-client-local.py
#!/usr/bin/env python3
"""
Ollama Client for Local Computer
আপনার local computer থেকে AWS Ollama ব্যবহার করার জন্য
"""

import requests
import json
import time
from typing import Optional, Dict, List

class OllamaClient:
    def __init__(self, base_url: str):
        """
        Initialize Ollama client
        
        Args:
            base_url: আপনার Ollama server-র URL (Example: http://54.123.45.67)
        """
        self.base_url = base_url.rstrip('/')
        self.api_url = f"{self.base_url}/api"
        
    def check_connection(self) -> bool:
        """Server connection check করুন"""
        try:
            response = requests.get(f"{self.api_url}/tags", timeout=5)
            return response.status_code == 200
        except:
            return False
    
    def list_models(self) -> List[str]:
        """Available models-র list পান"""
        try:
            response = requests.get(f"{self.api_url}/tags")
            data = response.json()
            return [model['name'] for model in data.get('models', [])]
        except Exception as e:
            print(f"Error: {e}")
            return []
    
    def generate(
        self, 
        prompt: str, 
        model: str = "llama3.2",
        stream: bool = False,
        temperature: float = 0.7
    ) -> str:
        """
        Text generate করুন
        
        Args:
            prompt: আপনার question/prompt
            model: Model name (default: llama3.2)
            stream: Streaming enable করবেন কিনা
            temperature: Creativity level (0.0 - 1.0)
        
        Returns:
            Generated text
        """
        try:
            payload = {
                "model": model,
                "prompt": prompt,
                "stream": stream,
                "options": {
                    "temperature": temperature
                }
            }
            
            response = requests.post(
                f"{self.api_url}/generate",
                json=payload,
                timeout=300  # 5 minutes timeout
            )
            
            if response.status_code == 200:
                data = response.json()
                return data.get('response', '')
            else:
                return f"Error: {response.status_code} - {response.text}"
                
        except Exception as e:
            return f"Error: {str(e)}"
    
    def chat(
        self,
        messages: List[Dict[str, str]],
        model: str = "llama3.2",
        stream: bool = False
    ) -> str:
        """
        Chat with conversation history
        
        Args:
            messages: List of messages [{"role": "user", "content": "..."}]
            model: Model name
            stream: Streaming enable করবেন কিনা
        
        Returns:
            AI response
        """
        try:
            payload = {
                "model": model,
                "messages": messages,
                "stream": stream
            }
            
            response = requests.post(
                f"{self.api_url}/chat",
                json=payload,
                timeout=300
            )
            
            if response.status_code == 200:
                data = response.json()
                return data.get('message', {}).get('content', '')
            else:
                return f"Error: {response.status_code}"
                
        except Exception as e:
            return f"Error: {str(e)}"


# Example usage
if __name__ == "__main__":
    # আপনার Ollama server-র URL দিন
    OLLAMA_URL = "http://54.123.45.67"  # আপনার EC2 IP দিন
    
    # Client initialize করুন
    client = OllamaClient(OLLAMA_URL)
    
    # Connection test করুন
    print("🔍 Testing connection...")
    if client.check_connection():
        print("✅ Connected to Ollama server!")
    else:
        print("❌ Cannot connect to Ollama server")
        exit(1)
    
    # Available models দেখুন
    print("\n📋 Available models:")
    models = client.list_models()
    for model in models:
        print(f"  - {model}")
    
    # Simple generation test
    print("\n💬 Testing text generation...")
    response = client.generate(
        prompt="বাংলাদেশের রাজধানী কোথায়?",
        model="llama3.2"
    )
    print(f"Response: {response}")
    
    # Chat test
    print("\n💬 Testing chat...")
    messages = [
        {"role": "user", "content": "What is Python?"},
        {"role": "assistant", "content": "Python is a programming language."},
        {"role": "user", "content": "What can I do with it?"}
    ]
    response = client.chat(messages, model="llama3.2")
    print(f"Response: {response}")
    
    print("\n✅ All tests completed!")
